Feature/atlasisoraycaster

RasmusHogslatt opened the pull request:
Summary:
* Adding one processor AtlasBoundary and a AtlasIsosurfaceComponent added to the current AtlasVolumeRaycaster processor.

AtlasBoundary:
Utilizes brushing and linking to convert a segmented volume into a UINT16 volume with voxel values in range [0, 3] and interpolation set to nearest.

AtlasIsosurfaceComponent (Added to AtlasVolumeRaycaster):
Adds option to render the segments produced by AtlasBoundary and toggle lighting on or off.

This has been tested on: Windows 10 / MSVC 143 / VS 2022 / QT 6.2.4 / Python 3.10

RasmusHogslatt committed:
Volume: Atlas boundary processor added

RasmusHogslatt committed:
BaseGL: Added atlas isosurface renderer processor

RasmusHogslatt committed:
BaseGL: Added atlasIsosurfaceRenderer

RasmusHogslatt committed:
Volume: Created workspace for testing

RasmusHogslatt committed:
BaseGL: Workspace created

RasmusHogslatt committed:
BaseGL: Added Atlas Isosurface raycasting component

RasmusHogslatt committed:
BaseGL: Added composite properties

RasmusHogslatt committed:
OpenGL: Sampler objects for gradients.

Work in progress

petersteneteg commented on modules/basegl/glsl/raycasting/atlasisosurfacerenderer.frag:
```diff
@@ -0,0 +1,130 @@
+    // Note that this computation is performed in world space
+    // f' = ( f(x+h)-f(x-h) ) / 2*volumeParams.worldSpaceGradientSpacing
+    return (cDs)/(2.0*volumeParams.worldSpaceGradientSpacing);
+}
```
Note: you also have to apply the scaling at the last step when dividing by the step size.

petersteneteg commented on modules/basegl/CMakeLists.txt:
```diff
@@ -167,6 +169,7 @@ set(SOURCE_FILES
     src/processors/meshrenderprocessorgl.cpp
     src/processors/multichannelraycaster.cpp
     src/processors/pointrenderer.cpp
+    src/processors/raycasting/atlasisosurfacerenderer.cpp
```
Is the any difference between this one and using the components? Is this just for debugging? I would prefer to only have the component one for maintenance reasons, unless there is a good reason to have. a separate one.

RasmusHogslatt replied:
Yes, I should have removed this one from the pull request. Atlasisosurfacerenderer is purely for debugging purposes.

petersteneteg commented on modules/basegl/include/modules/basegl/processors/raycasting/atlasvolumeraycaster.h:
```diff
@@ -93,6 +95,7 @@ class IVW_MODULE_BASEGL_API AtlasVolumeRaycaster : public VolumeRaycasterBase {
     PositionIndicatorComponent positionIndicator_;
     SampleTransformComponent sampleTransform_;
     AtlasComponent atlas_;
+    AtlasIsosurfaceComponent atlasIso_;
```
Question? is this related to isosurfaces? Is this not for showing the segment boundaries? that is probably not an iso surface right? Maybe rename to SegementSurfaceComponent or RegionSurfaceComponent?

RasmusHogslatt replied:
Yes, that sounds better.

petersteneteg commented on modules/basegl/include/modules/basegl/shadercomponents/atlasisosurfacecomponent.h:
```diff
@@ -0,0 +1,71 @@
+ * Adds an option to render segments in range [0, 3] of type UINT, and render these
+ * with or without lighting. See the Atlas Boundary processor for such data.
+ */
+class IVW_MODULE_BASEGL_API AtlasIsosurfaceComponent : public ShaderComponent {
```
Again rename...

petersteneteg commented on modules/basegl/include/modules/basegl/shadercomponents/atlasisosurfacecomponent.h:
```diff
@@ -0,0 +1,71 @@
+    SelectionColorProperty showHighlighted_;
+    SelectionColorProperty showSelected_;
+    SelectionColorProperty showFiltered_;
+    SamplerObject nearestSampler_;
```
You only need the one linear one or?

RasmusHogslatt replied:
I actually need both unless I got something wrong. Creating only one linear would automatically take the "spot" of the regular texture sampler in the shader. So, as long as a samplerobject exists for a texture unit, that sampler object will take its spot in the shader. Thus, having both as in the current implementation, the textures parameters will be overwritten by nearestSampler, and linearSampler will get the next spot. The shader thus becomes
`uniform sampler3D nearestSampler; // Without samplerobjects, this would have had the texture's parameters
uniform sampler3D linearSampler;`

petersteneteg replied:
I think this is since you don't assign specific texture units for the sampler but just pass in 0 as texture unit. You should assign a dedicated texture unit for the sampler then I should not mess with existing ones. You also need to assign the uniform associated with that sampler to that new texture unit

petersteneteg commented on modules/basegl/src/processors/raycasting/atlasisosurfacerenderer.cpp:
```diff
@@ -0,0 +1,143 @@
+    utilgl::bindAndSetUniforms(shader_, units, volume, "volume");
+    utilgl::bindAndSetUniforms(shader_, units, entryPort_, ImageType::ColorDepthPicking);
+    utilgl::bindAndSetUniforms(shader_, units, exitPort_, ImageType::ColorDepth);
+
```
You should bind the SamplerObject to a unit here, and bind the altas to the same unit. Maybe that inviwo::SamplerObject should handle that automatically

petersteneteg commented on modules/basegl/include/modules/basegl/shadercomponents/atlasisosurfacecomponent.h:
```diff
@@ -0,0 +1,71 @@
+    virtual std::vector<Segment> getSegments() override;
+
+private:
+    VolumeInport volume_;
```
Should have a name like selectionAtlasInport, Or something descriptive...

petersteneteg commented on modules/basegl/src/processors/raycasting/atlasvolumeraycaster.cpp:
```diff
@@ -58,17 +58,19 @@ AtlasVolumeRaycaster::AtlasVolumeRaycaster(std::string_view identifier,
     , sampleTransform_{}
-    , atlas_{this, "color", &time_} {
+    , atlas_{this, "color", &time_}
+    , atlasIso_{"atlasiso"} {
```
Not sure if this should be in the regular altas renders, Has to be optional In that case I think.

RasmusHogslatt replied:
I'm a bit confused. Suggestion marks removal of the atlas_, which is not what I have worked on. Could you explain what should be an option here?

petersteneteg replied:
No the altas is still there It's just that the "{" at then end is remove and moved to the end of the atlasIso line.

petersteneteg commented on modules/basegl/src/shadercomponents/atlasisosurfacecomponent.cpp:
```diff
@@ -0,0 +1,178 @@
+
+AtlasIsosurfaceComponent::AtlasIsosurfaceComponent(std::string_view volume)
+    : ShaderComponent{}
+    , volume_{"atlasboundary"}
```
You should make this port optional

petersteneteg commented on modules/basegl/src/shadercomponents/atlasisosurfacecomponent.cpp:
```diff
@@ -0,0 +1,178 @@
+    shader.setUniform("showHighlighted", showHighlighted_.getBoolProperty()->get());
+    shader.setUniform("filteredColor", showFiltered_.getColor());
+    shader.setUniform("showFiltered", showFiltered_.getBoolProperty()->get());
+
```
You should bind your samplerObject to a texture unit here.

petersteneteg commented on modules/basegl/src/shadercomponents/atlasisosurfacecomponent.cpp:
```diff
@@ -0,0 +1,178 @@
+    }*/
+}
+
+std::vector<Property*> AtlasIsosurfaceComponent::getProperties() { return {&useAtlasBoundary_}; }
```
add the other properties

RasmusHogslatt replied:
useAtlasBoundary_ is a boolCompositeProperty with the other properties added to it. I tried adding everything here before, but then I got duplicates of all properties. Martin suggested to put all properties in a composite one for clarity.

petersteneteg replied:
ahh ok

petersteneteg commented on modules/basegl/src/shadercomponents/atlasisosurfacecomponent.cpp:
```diff
@@ -0,0 +1,178 @@
+if(!useAtlasBoundary)
+{
+    // Do nothing
+}
```
We probably don't what this as a runtime option, we should probably recompile the shader when this changes...
@martinfalk are the compiler smart enough to understand that this value will be the same for all voxels and not have to do both paths?

petersteneteg commented on modules/basegl/src/shadercomponents/raycastingcomponent.cpp:
```diff
@@ -70,6 +70,7 @@ RaycastingComponent::RaycastingComponent(std::string_view volume, IsoTFProperty&
 std::string_view RaycastingComponent::getName() const { return raycasting_.getIdentifier(); }
 
 void RaycastingComponent::process(Shader& shader, TextureUnitContainer&) {
+    shader.setUniform("gradientOffsetSpacing", raycasting_.gradientSpacingOffset_.get());
```
I'm not sure we need this setting in the general case. I have never had any reason to use it before.

RasmusHogslatt replied:
Okay, do we still want this option for this processor? If so, should I copy gradientCentralDiff and apply scaling to it, or should there be an overloaded version for all of the gradient calculations? Next comment clarified this.

petersteneteg replied:
Guess we can have a specific one for the surface here.

petersteneteg replied:
you can just create a copy  of the volumeParams in the shader and update the textureSpaceGradientSpacing to the appropriate value

petersteneteg commented on modules/opengl/include/modules/opengl/texture/samplerobject.h:
```diff
@@ -0,0 +1,64 @@
+
+/**
+ * Sampler objects store parameters that are not specific to textures. Thus, they store filter modes, wrapping mode but not texture width, height, swizzlemask etc. Binding a sampler object to a texture unit takes that textures uniform location in the shader, and multiple different sampler objects can be assigned to the same texture unit without having to copy the texture data.
+ */
```
Linewrap

petersteneteg commented on modules/opengl/include/modules/opengl/texture/samplerobject.h:
```diff
@@ -0,0 +1,64 @@
+class IVW_MODULE_OPENGL_API SamplerObject {
+public:
+    SamplerObject();
+    virtual ~SamplerObject();
```
Should not be virtual

petersteneteg commented on modules/opengl/src/texture/samplerobject.cpp:
```diff
@@ -0,0 +1,82 @@
+
+namespace inviwo {
+
+SamplerObject::SamplerObject() : currentlyBoundTextureUnit_{0} { glGenSamplers(1, &id_); }
```
Should add parameter options to the constructor

petersteneteg commented on modules/opengl/src/texture/samplerobject.cpp:
```diff
@@ -0,0 +1,82 @@
+
+void SamplerObject::bind(GLuint textureUnitNumber) {
+    glBindSampler(textureUnitNumber, id_);
+    currentlyBoundTextureUnit_ = textureUnitNumber;
```
why is currentlyBoundTextureUnit_ needed? can it be removed?

RasmusHogslatt replied:
Not needed as of right now, but I think it might be useful for debugging to figure out which unit you are working on. Mostly for debugging though. Thus it can be removed. Though, every time you bind, you need to specify which unit to bind to. If you have to bind lots of times, it might be nicer to have what unit it's bound to rather than setting it manually in every bind call.

petersteneteg replied:
In out current approach we bind all texture unit every time we change shader anyway so I don't think that is a problem.. There is a limited number of texture units so you have to take care of tracking and reusing them

petersteneteg commented on modules/opengl/include/modules/opengl/texture/samplerobject.h:
```diff
@@ -0,0 +1,64 @@
+    void bind(GLuint textureUnitNumber);
+    void setMinFilterMode(GLenum type);
+    void setMagFilterMode(GLenum type);
+    void setFilterModeAll(GLenum type);
```
should add overloads for https://github.com/inviwo/inviwo/blob/614c8793c8b1ac9bdd4444a90d97217d9201a3fb/include/inviwo/core/datastructures/image/imagetypes.h#L49-L54

petersteneteg commented on modules/opengl/src/texture/samplerobject.cpp:
```diff
@@ -0,0 +1,82 @@
+SamplerObject::~SamplerObject() {
+    /*if (id_ != 0) {
+        glDeleteSamplers(1, &id_);
+    }*/ // Causes "unable to upload renderui.vert"
```
If this fails we are doing something wrong which need to be fixed...

RasmusHogslatt replied:
Yes, maybe we can have a look at it during next meeting.

petersteneteg replied:
yes

petersteneteg commented on modules/volume/src/processors/atlasboundary.cpp:
```diff
@@ -0,0 +1,97 @@
+const ProcessorInfo AtlasBoundary::processorInfo_{
+    "org.inviwo.atlasboundary",  // Class identifier
+    "Atlas Boundary",            // Display name
+    "Volume Rendering",          // Category
```
Don't think Rendering is the right category

RasmusHogslatt replied:
True, will check for a more suitable catergory.

RasmusHogslatt replied:
Changed to "Volume Operation", same as for voronoi segmentation.

petersteneteg commented on modules/volume/src/processors/atlasboundary.cpp:
```diff
@@ -0,0 +1,97 @@
+        volume_ = std::shared_ptr<Volume>(volume->clone());
+        util::remap(*volume_, sourceIndices, destinationIndices, 0, true);
+        // Set volume properties
+        volume_->setInterpolation(InterpolationType::Nearest);
```
remove?

petersteneteg commented on modules/volume/src/processors/atlasboundary.cpp:
```diff
@@ -0,0 +1,97 @@
+        destinationIndices.insert(destinationIndices.end(), filteredDestination.begin(),
+                                  filteredDestination.end());
+        destinationIndices.insert(destinationIndices.end(), highlightedDestination.begin(),
+                                  highlightedDestination.end());
```
Could use std::fill_n tree time here instead of creating 3 temp vectors...

petersteneteg commented on modules/volume/src/processors/atlasboundary.cpp:
```diff
@@ -0,0 +1,97 @@
+
+        // Append to B&L to vectors
+        std::vector<int> sourceIndices;
+        std::vector<int> destinationIndices;
```
You might want to keep this 2 vectors around as members to avoid having to create them each time...

RasmusHogslatt committed:
Update modules/basegl/include/modules/basegl/shadercomponents/atlasisosurfacecomponent.h

Co-authored-by: Peter Steneteg <peter@steneteg.se>

RasmusHogslatt committed:
OpenGL: SamplerObjects

RasmusHogslatt committed:
BaseGL: Cleanup

RasmusHogslatt committed:
BaseGL: B&L rendering + Samplerobjects

RasmusHogslatt committed:
Volume: Update regression test

Feature added caused change setting.

RasmusHogslatt committed:
BaseGL: Regression fix

BaseGL: Formatting

BaseGL: Regression

BaseGL: Format fixes

BaseGL: Format

BaseGL: Format

BaseGL: Formatting

RasmusHogslatt committed:
BaseGL: Updated packages

petersteneteg commented on modules/basegl/include/modules/basegl/shadercomponents/segmentsurfacecomponent.h:
```diff
@@ -0,0 +1,70 @@
+    SelectionColorProperty showHighlighted_;
+    SelectionColorProperty showSelected_;
+    SelectionColorProperty showFiltered_;
+    SamplerObject nearestSampler_;
```
This one is not used anymore I think and should be removed

petersteneteg commented on modules/basegl/src/shadercomponents/segmentsurfacecomponent.cpp:
```diff
@@ -0,0 +1,154 @@
+                                    showHighlighted_, showFiltered_, showSelected_);
+
+    nearestSampler_.setFilterModeAll(InterpolationType::Nearest);
+    nearestSampler_.setWrapModeAll(Wrapping::Repeat);
```
Remove

petersteneteg commented on modules/basegl/src/shadercomponents/segmentsurfacecomponent.cpp:
```diff
@@ -0,0 +1,154 @@
+    nearestSampler_.setFilterModeAll(InterpolationType::Nearest);
+    nearestSampler_.setWrapModeAll(Wrapping::Repeat);
+    linearSampler_.setFilterModeAll(InterpolationType::Linear);
+    linearSampler_.setWrapModeAll(Wrapping::Repeat);
```
Should set these as constructor arguments above

RasmusHogslatt replied:
Add filter and wrap mode to this `SegmentSurfaceComponent(VolumeInport& atlas, InterpolationType type, Wrapping wrap)` or add these line to initializer list? Assume the former of the two.

petersteneteg commented on modules/basegl/src/shadercomponents/segmentsurfacecomponent.cpp:
```diff
@@ -0,0 +1,154 @@
+        boundaryColor.rgb = APPLY_LIGHTING(lighting, boundaryColor.rgb, boundaryColor.rgb,
+            vec3(1.0), worldSpacePosition, -atlasGradient, cameraDir);  
+    }   
+    #endif
```
The opening "{" should be on the same line, according to our style guide

petersteneteg commented on modules/opengl/include/modules/opengl/texture/samplerobject.h:
```diff
@@ -0,0 +1,79 @@
+ */
+class IVW_MODULE_OPENGL_API SamplerObject {
+public:
+    SamplerObject();
```
Add constructor overloads for the InterpolationType and Wrapping

petersteneteg commented on modules/opengl/include/modules/opengl/texture/samplerobject.h:
```diff
@@ -0,0 +1,79 @@
+
+private:
+    GLuint id_;
+    GLuint currentlyBoundTextureUnit_;
```
We can remove currentlyBoundTextureUnit_ now.

petersteneteg commented on modules/opengl/include/modules/opengl/texture/samplerobject.h:
```diff
@@ -0,0 +1,79 @@
+    SamplerObject& operator=(SamplerObject&&) noexcept;
+    ~SamplerObject();
+
+    /// <summary>
```
We don't use this style (html) for comments

petersteneteg commented on modules/opengl/include/modules/opengl/texture/samplerobject.h:
```diff
@@ -0,0 +1,79 @@
+ * Sampler objects store parameters that are not specific to textures. Thus, they store filter
+ * modes, wrapping mode but not texture width, height, swizzlemask etc. Binding a sampler object to
+ * a texture unit takes a texture's uniform location in the shader, and multiple different sampler
+ * objects can be assigned to the same texture unit without having to copy the texture data.
```
A texture unit should only have one or zero sampler object and one texture bound at any time. 
Might att a link to the bind(sampler object function) in texture unit

petersteneteg reviewed (changes requested):
Looking nice now, found a few minor issues to address before merging.

RasmusHogslatt committed:
BaseGL: Format + Overloads

RasmusHogslatt committed:
BaseGL: Format

RasmusHogslatt committed:
BaseGL: Constructor changes

petersteneteg committed:
OpenGL: missing include

Inviwo-Team committed:
Jenkins: Format fixes

Inviwo-Team committed:
Jenkins: Format fixes
